deepchem/models/torch_models/layers.py

```diff
@@ -2796,3 +2796,147 @@ def forward(
+            The length of the pair features vector.
+        n_hidden: int, optional
+            number of hidden units in the passing phase
+        init: str, optional
+            Initialization function to be used in the message passing layer.
+        """
+
+        super(EdgeNetwork, self).__init__(**kwargs)
+        self.n_pair_features: int = n_pair_features
+        self.n_hidden: int = n_hidden
+        self.init: str = init
+
+        init_func: Callable = getattr(initializers, self.init)
+        self.W: torch.Tensor = init_func(
+            torch.empty([self.n_pair_features, self.n_hidden * self.n_hidden]))
+        self.b: torch.Tensor = torch.zeros((self.n_hidden * self.n_hidden,))
+        self.built: bool = True
+
+    def __repr__(self) -> str:
+        return (
+            f'{self.__class__.__name__}(n_pair_features:{self.n_pair_features},n_hidden:{self.n_hidden},init:{self.init})'
+        )
+
+    def forward(self, inputs: List[torch.Tensor]) -> torch.Tensor:
+        """
+        Parameters
+        ----------
+        inputs: List[torch.Tensor]
+            The length of atom_to_pair should be same as n_pair_features.
+
+        Returns
+        -------
+        result: torch.Tensor
+            Tensor containing the mapping of the edge vector to a d × d matrix, where d denotes the dimension of the internal hidden representation of each node in the graph.
+        """
+        pair_features: torch.Tensor
+        atom_features: torch.Tensor
+        atom_to_pair: torch.Tensor
+        pair_features, atom_features, atom_to_pair = inputs
+
+        def unsorted_segment_sum(data, segment_ids, num_segments):
+            """Computes the sum along segments of a tensor. Analogous to tf.unsorted_segment_sum.
+
+            Parameters
+            ----------
+            data: A tensor whose segments are to be summed.
+            segment_ids: The segment indices tensor.
+            num_segments: The number of segments.
+
+            Returns
+            -------
+            tensor: torch.Tensor
+            """
+            assert all([i in data.shape for i in segment_ids.shape
+                       ]), "segment_ids.shape should be a prefix of data.shape"
+
+            # segment_ids is a 1-D tensor repeat it to have the same shape as data
+            if len(segment_ids.shape) == 1:
+                s = torch.prod(torch.tensor(data.shape[1:])).long()
+                segment_ids = segment_ids.repeat_interleave(s).view(
+                    segment_ids.shape[0], *data.shape[1:])
+
+            assert data.shape == segment_ids.shape, "data.shape and segment_ids.shape should be equal"
+
+            shape = [num_segments] + list(data.shape[1:])
+            tensor = torch.zeros(*shape).scatter_add(0, segment_ids,
+                                                     data.float())
+            tensor = tensor.type(data.dtype)
+            return tensor
+
+        def segment_sum(data, segment_ids):
```

Comment:
add these functions in the `deepchem/utils/pytorch_utils.py`  file so that others working on porting the layers don't reimplement the functions